### PR TITLE
New Olivia

### PR DESCRIPTION
**Pull request by gondiaz**
This is Olivia modified:
-Change in baseline calculation for SiPM
-Add Adam plots
-Add scale for the histogram class

### CODE REVIEW

**Comment by andLaing** [comment]
Some tests seem to be failing because of the absence of certain fixtures: 'ICDATADIR' and 'output_tmpdir' for example. Suspect that the conftest.py file hasn't been uploaded along with the changes.

**Comment by gondiaz** [comment]
I have missed to upload the conftest. Try now.

**Review by andLaing — COMMENTED** [review]
There's still a couple of test fails in monitor_functions_test:

test_fill_rwf_histos: seems to be test data related, update to the latest version of IC and check.
test_fill_kdst_histos: 

> TypeError: create_histomanager_from_dicts() missing 1 required positional argument: 'histoscales_dict'

One of the functions hasn't been updated fully by the looks of things.

**Comment by andLaing** [comment]
`for k, v in config_dict.items():
        if   "_bins"   in k: var_bins  [k.replace("_bins"  , "")] = [np.linspace(v[0], v[1], v[2] + 1)]
        elif "_labels" in k: var_labels[k.replace("_labels", "")] = v
        elif "_scales" in k: var_scales[k.replace("_scales", "")] = v

    del var_bins  ["PMTs_AdamPlot"]
    del var_labels["PMTs_AdamPlot"]
    del var_scales["PMTs_AdamPlot"]

    n_PMTs = config_dict["n_PMTs"]
    v      = config_dict["PMTs_AdamPlot_bins"]
    for i in range(0, int(n_PMTs)):
        var_bins  [f"PMT{i}_AdamPlot"] = [np.linspace(v[0], v[1], v[2] + 1)]
        var_labels[f"PMT{i}_AdamPlot"] = [f"PMT{i}_AdamPlot (ADC)"]
        var_scales[f"PMT{i}_AdamPlot"] = config_dict["PMTs_AdamPlot_scales"]`

It seems inefficient to add in the the 'AdamPlots' then remove them to add in a different way. Maybe better to skip over them in the first step?? Also, What is an AdamPlot?? Give them a descriptive name, please

**Comment by andLaing** [comment]
`if   sensor_type == "SiPM":`

All occurrences of this type of comparison should be replaced by Enum.

**Comment by andLaing** [comment]
In pmap_bins there's another del after adding values, also the function is really difficult to read, maybe needs a bit of a tidy

**Comment by andLaing** [comment]
In monitor_functions lines 184 -- 188:
`for in_file in glob.glob(in_path):
        pmaps = load_pmaps(in_file)
        for ti, pi in enumerate(pmaps):
            var = fill_pmap_var(pmaps[pi], SiPM_db)
            histo_manager.fill_histograms(var)`

there is no need for the enumerate nor to have access to the keys. Use:
`for pmap in pmaps.values():
           var = fill_pmap_var(pmap, SiPM_db)`

or similar

File: olivia/config/olivia_config.conf
Comment (andLaing) [inline olivia/config/olivia_config.conf]: the multiple data_type definitions imply that multiple types could be analysed in the same job but the olivia_script.py executable doesn't seem to have this functionality.

If multiple data_type definitions do nothing make separate example configs for each case.
Also, probably put some of the test data as the example inputs rather than absolute paths to your instalation.

File: olivia/hist_io.py
Comment (andLaing) [inline olivia/hist_io.py:8]: The alignment is a bit off and on in all functions, once things are ready I'd recommend a bit of a tidy in general

File: olivia/monitor_functions.py
Comment (andLaing) [inline olivia/monitor_functions.py]: As mentioned in an earlier comment, I'd prefer to skip over rather than add, delete, add

File: olivia/monitor_functions.py
Comment (andLaing) [inline olivia/monitor_functions.py:270]: Are 'AdamPlots' just raw waveforms? If so, just call them that. If not, give them a descriptive name

File: olivia/olivia_script.py
Comment (andLaing) [inline olivia/olivia_script.py]: Probably better changed to an Enum

File: olivia/config/olivia_config.conf
Comment (andLaing) [reply olivia/config/olivia_config.conf]: For the configs you should use $OLIVIADIR/.... so it's set correctly for all users.

**Comment by andLaing** [comment]
We're missing an example config for kdst mode too

**Comment by Aretno** [comment]
Regarding the failing test, we've been looking into this and it seems to be caused by incompatibilities between latest versions of pytables and numpy (https://github.com/PyTables/PyTables/issues/719#issuecomment-455612656)

From what I can gather, numpy latest version added some extra protection to avoid writing where it should not (basically not allowing ndarray to be writable) and pytables is setting it to be writable and causes the problem. 

So here we have a couple of options, either use numpy 1.15 (the problem starts with numpy 1.16) or wait till pytables releases the fix (which it seems they have written but not made a release). Since this is a bit tricky I wonder if we should bring this question to bosses/superexperts.

EDIT: Actually, my IC seems to be at numpy 1.15 (and so it appears in the manage.sh of IC); Gonzalo reproduced failures at 1.16... Andrew which numpy version do you have?

**Comment by andLaing** [comment]
> EDIT: Actually, my IC seems to be at numpy 1.15 (and so it appears in the manage.sh of IC); Gonzalo reproduced failures at 1.16... Andrew which numpy version do you have?

conda list gives me:
numpy                     1.15.2           py37h6a91979_1

**Comment by Aretno** [comment]
That's weird, what is the exact error you're getting in the test_fill_rwf_histos? Maybe it's not the same as the one that Gonzalo sees and we are comparing different things. For him, in addition, it also fails in test_fill-pmap_histos...

In any case, additionally, please check the numpy version with:

import numpy
numpy.version.version

Gonzalo has the same one in conda list but when importing the module, it claims to be 1.16.2 and thus the error.

**Comment by gondiaz** [comment]
> There's still a couple of test fails in monitor_functions_test:
> 
> test_fill_rwf_histos: seems to be test data related, update to the latest version of IC and check.
> test_fill_kdst_histos:
> 
> > TypeError: create_histomanager_from_dicts() missing 1 required positional argument: 'histoscales_dict'
> 
> One of the functions hasn't been updated fully by the looks of things.

**Comment by andLaing** [comment]
> import numpy
numpy.version.version

output: '1.15.2'

> what is the exact error you're getting in the test_fill_rwf_histos?

assert False
E            +  where False = <function allclose at 0x10e97e9d8>(array([1.786e+03, 0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00, 2.000e+00,\n       1.000e+00, 0.000e+00, 0.000e+00, 0.000...  0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,\n       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00]), array([1786.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n          0.,    0.,    0.,    0.,    0.,    0.,...   0.,    0.,    0.,    0.,    0.,\n          0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n          0.]))
E            +    where <function allclose at 0x10e97e9d8> = np.allclose
E            +    and   array([1.786e+03, 0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00, 2.000e+00,\n       1.000e+00, 0.000e+00, 0.000e+00, 0.000...  0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,\n       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00]) = <olivia.histos.Histogram object at 0x1c1f6ffe80>.data
E            +    and   array([1786.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n          0.,    0.,    0.,    0.,    0.,    0.,...   0.,    0.,    0.,    0.,    0.,\n          0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n          0.]) = <olivia.histos.Histogram object at 0x1c223ef080>.data

olivia/monitor_functions_test.py:409: AssertionError

**Comment by andLaing** [comment]
Why has this been closed?

A

**Comment by carmenromo** [comment]
Uops, it was an error

**Comment by carmenromo** [comment]
I am trying to solve it with JM

**Comment by carmenromo** [comment]
Andrew, this PR is ready to review

**Review by andLaing — COMMENTED** [review]
Still missing a .json for kdst plots in the most recent version

File: olivia/monitor_functions.py
Comment (gondiaz) [reply olivia/monitor_functions.py]: This was changed, are you sure you are in the correct branch and commit?

File: olivia/monitor_functions.py
Comment (gondiaz) [reply olivia/monitor_functions.py:270]: Same as above. AdamPlot names were changed.

File: olivia/config/olivia_config.conf
Comment (gondiaz) [reply olivia/config/olivia_config.conf]: That is true that is misleading, but it just runs for the data_type in the last line. I had this to do not write the parameters every time I run, I just copied-paste above or below.

Comment (gondiaz) [reply olivia/config/olivia_config.conf]: Ander said not to include kdsts for the moment. In fact, all the kdsts related code is commented now

File: olivia/monitor_functions.py
Comment (andLaing) [reply olivia/monitor_functions.py]: I hit resolve, I assumed that closed the issue. Sorry for the confusion

File: olivia/monitor_functions.py
Comment (andLaing) [reply olivia/monitor_functions.py:270]: Again, resolve... maybe it doesn't do what I thought

File: olivia/config/olivia_config.conf
Comment (andLaing) [reply olivia/config/olivia_config.conf]: OK, since this is an example it should be runable or at least clear.

Make the simple changes and put a comment about the other possibilities.

File: olivia/histos.py
Comment (andLaing) [inline olivia/histos.py:8]: It's not entirely clear to me what 'scale' does.

Is it related to the normalisation? Is it log vs. linear. I think a more descriptive name needs to be found

File: olivia/olivia_script.py
Comment (andLaing) [inline olivia/olivia_script.py]: If the kdst option is not ready this shouldn't be there.

**Comment by andLaing** [comment]
All tests seem to pass.

Just a few comments to deal with and a bit of a tidy. I'll check the outputs but you already showed them to me separately I think.

Should the drawing part be here or is it separate?

File: olivia/histos.py
Comment (gondiaz) [reply olivia/histos.py:8]: scale is a Histogram atribute for plotting purpuses, it says how the Histogram y axis should be plotted and do not affect the Histogram class by itself. This behaviour is the same as title and labels atributes, which are just for identification and plotting purpuses. Scale should have one of the matplotlib allowed scales: linear, log, symlog, logit.

Note that the code is retrocompatible since this scale is set automatically to linear when importing old olivia histograms in the `get_histograms_from_file` function in hist_io module.

**Comment by gondiaz** [comment]
The outputs I have shown are temporarily here:
https://github.com/gondiaz/Olivia/blob/prod_example/plot_histo.ipynb

If with the drawing part you mean the histogram_plot_functions module, I think its more confortable to be here in case we want to change something in the plotting, such as colormappings, etc..

There is another issue that is to set the ranges in the .json, that there are not still perfect (SiPM baseline is not properly set because a strange problem in one run), but I don't know if this should be set here or whoever uses Olivia sets them as wished.

**Comment by andLaing** [comment]
Where are we on this?

**Comment by gondiaz** [comment]
We have to review the ranges in the .json. Once done, I think it will be finished. I am having problems in my computer, so once I have a new one I could do that. But maybe @carmenromo is available to do that

**Comment by carmenromo** [comment]
Done!

**Comment by andLaing** [comment]
I've pushed some cosmetic changes. They are by no means complete or perfect so have a look and add to them.

I still think the example config should be split into two. A minimum change should be to remove absolute paths pointing to personal file structures.

**Comment by andLaing** [comment]
Good change but we need to remove the /home/gonzalo etc

**Comment by carmenromo** [comment]
You are right, sorry!

**Comment by carmenromo** [comment]
Why is the code commented from line 287 in the file monitor_functions.py?

**Comment by gondiaz** [comment]
> Why is the code commented from line 287 in the file monitor_functions.py?

Because these are the kdsts related functions

**Comment by carmenromo** [comment]
I wonder if maybe we should use Enum for ptype (monitor_functions) as well.

**Comment by andLaing** [comment]
where's ptype in monitor functions? I don't see it in the diff

**Comment by carmenromo** [comment]
In fill_pmap_var_1d and fill_pmap_var_2d functions. It is a string

**Comment by andLaing** [comment]
It's used as a string in many places though you could change the way the titles are set, I suppose. Up to the authors.

**Comment by andLaing** [comment]
Gente!!!

When are we going to approve this?

**Comment by gondiaz** [comment]
Hi Andrew, I think we are just waiting for your approval.

**Comment by andLaing** [comment]
I thought there were a couple of my requests still outstanding. I'll have a bit of a look later this afternoon and see what the status is.

File: olivia/histos.py
Comment (andLaing) [reply olivia/histos.py:8]: It's still not clear what this does. Is it whether or not you normalise? It's a float so it's not just a decision between log or linear scale.

**Comment by andLaing** [comment]
> It's used as a string in many places though you could change the way the titles are set, I suppose. Up to the authors.

Was there a decision on this from @carmenromo and/or @gondiaz or does @Aretno overrule?

**Comment by carmenromo** [comment]
> > It's used as a string in many places though you could change the way the titles are set, I suppose. Up to the authors.
> 
> Was there a decision on this from @carmenromo and/or @gondiaz or does @Aretno overrule?

Sorry I din't do anything because I thought it was going to be a bigger change in the code that would affect IC, so I was not sure....

**Review by andLaing — APPROVED** [review]
The work is very complete with just a few questions that are superficial could be put in a different PR if the authors prefer.

Approved and ready for merging

File: olivia/histos.py
Comment (gondiaz) [reply olivia/histos.py:8]: I think I don't get your point. **scale** is not a float, it is a string with the posible matplotlib allowed scales linear, log, symlog, logit. The normalization is set to True in histogram_plot_functions.py. 

In fact this is something that I talk with @Aretno to modify because this normalizes the histogram integral, not the bin values. But finally we let this as it was.

**Comment by carmenromo** [comment]
I will merge

File: olivia/histos.py
Comment (andLaing) [reply olivia/histos.py:8]: In the docstring it is described as a float and since it's not obvious where it's used I couldn't contradict this.
Do a PR correcting the error, please.

**Pull request merged**
